OSDOCS#10702: Docs for Multiarch Tuning Operator (#76628)
openshift/openshift-docs: post_installation_configuration/configuring-multi-arch-compute-machines/multiarch-tuning-operator.adoc

:_mod-docs-content-type: ASSEMBLY
:context: multi-architecture-tuning-operator
[id="multiarch-tuning-operator"]
= Managing a multi-architecture cluster by using the Multiarch Tuning Operator
include::_attributes/common-attributes.adoc[]

toc::[]

The Multiarch Tuning Operator enhances operational efficiency within multi-architecture clusters by facilitating architecture-aware workload scheduling. It schedules the workloads by considering the specific architectures of each compute node within the cluster. 

You can use the Multiarch Tuning Operator to perform the following actions: 

> * Managing clusters with multi-architecture compute machines.

Reviewer: "Managing workload placement on clusters with multi-architecture compute nodes"

Author: How about saying: "Scheduling workloads on clusters with multi-architecture compute machines"

> * Migrating from a single-architecture cluster to a multi-architecture cluster.

Reviewer: this can be removed and added as a detail below that this operator improves the migration experience of workloads from a single->multi arch compute cluster

Author: Hi @Prashanth684 
I have updated the intro section. Could you PTAL?
Thanks!

:_mod-docs-content-type: ASSEMBLY
:context: multi-architecture-tuning-operator
[id="multiarch-tuning-operator"]
= Managing workloads on multi-architecture clusters by using the Multiarch Tuning Operator
include::_attributes/common-attributes.adoc[]

toc::[]

[IMPORTANT]
====
The Multiarch Tuning Operator is a Technology Preview feature only. Technology Preview features are not supported with Red{nbsp}Hat production service level agreements (SLAs) and might not be functionally complete. Red{nbsp}Hat does not recommend using them in production. These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.
For more information about the support scope of Red{nbsp}Hat Technology Preview features, see link:https://access.redhat.com/support/offerings/techpreview/[Technology Preview Features Support Scope].
====

The Multiarch Tuning Operator enhances the operational experience within multi-architecture clusters, and single-architecture clusters that are migrating to a multi-architecture compute configuration. 

> This Operator uses the `ClusterPodPlacementConfig` custom resource (CR) to support architecture-aware workload scheduling.

Reviewer 2: ```suggestion
This Operator implements the `ClusterPodPlacementConfig` custom resource (CR) to support architecture-aware workload scheduling.
```

Author: Updated as per the suggestion.

:_mod-docs-content-type: ASSEMBLY
:context: multi-architecture-tuning-operator
[id="multiarch-tuning-operator"]
= Managing workloads on multi-architecture clusters by using the Multiarch Tuning Operator
include::_attributes/common-attributes.adoc[]

toc::[]

[IMPORTANT]
====
The Multiarch Tuning Operator is a Technology Preview feature only. Technology Preview features are not supported with Red{nbsp}Hat production service level agreements (SLAs) and might not be functionally complete. Red{nbsp}Hat does not recommend using them in production. These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.
For more information about the support scope of Red{nbsp}Hat Technology Preview features, see link:https://access.redhat.com/support/offerings/techpreview/[Technology Preview Features Support Scope].
====

> The Multiarch Tuning Operator enhances operational efficiency within multi-architecture clusters by facilitating architecture-aware workload scheduling. It schedules the workloads on clusters with multi-architecture compute machines by considering the specific architectures of each compute node within the cluster. It also simplifies the process of migrating your cluster with single-architecture compute machines to a cluster with multi-architecture compute machines.
> 
> The Multiarch Tuning Operator facilitates architecture-aware workload scheduling by using the `PodPlacementConfig` custom resource (CR).
> 
> The `PodPlacementConfig` CR automates the configuration of predicates based on the `kubernetes.io/arch` label that is found in the `nodeAffinity` field in the pod's specification. It analyzes the container images within each pod to identify supported architectures, ensuring compatibility and efficiency during pod deployment.
> 
> The `PodPlacementConfig` CR manages the following objects:
> 
> * `PodPlacementController`
> * `PodPlacementWebhook`
> 
> When you create a pod, the `PodPlacementWebhook` object adds the `multiarch.openshift.io/scheduling-gate` scheduling gate. This delays the scheduling of the pod until the `PodPlacementController` object completes the following actions:
> 
> . Computes a predicate for the `kubernetes.io/arch` label. 
> . Integrate the predicate as a node affinity requirement in the pod's specification.
> . Remove the scheduling gate.

Reviewer 2: ```suggestion
The Multiarch Tuning Operator aims to enhance the operational experience within multi-architecture clusters and single-architecture clusters migrating to a multi-architecture compute configuration.

The Multiarch Tuning Operator implements the `ClusterPodPlacementConfig` cluster-scoped Custom Resource (CR) to support the architecture-aware workload scheduling.

The operand managed via the `ClusterPodPlacementConfig` automates the configuration of scheduling predicates based on the `kubernetes.io/arch` label set by the Kubelet in the Node resources. For any newly created Pod, it sets the `nodeAffinity` field in the pod's specification by computing the subset of architectures that the pod can support.

Before a pod enters the scheduling cycle, the operand adds the `multiarch.openshift.io/scheduling-gate` scheduling gate that prevents the scheduling of the pod until the operand completes the following actions:

. Computes a predicate for the `kubernetes.io/arch` label. 
. Integrate the predicate as a node affinity requirement in the pod's specification.
. Remove the scheduling gate.
```

:_mod-docs-content-type: ASSEMBLY

> :context: multi-architecture-tuning-operator

Reviewer 3: ```suggestion
:context: multiarch-tuning-operator
```
I like to make the context tag and ID the same (easier later for xrefs)

[id="multiarch-tuning-operator"]
= Managing workloads on multi-architecture clusters by using the Multiarch Tuning Operator
include::_attributes/common-attributes.adoc[]

toc::[]

[IMPORTANT]
====
The Multiarch Tuning Operator is a Technology Preview feature only. Technology Preview features are not supported with Red{nbsp}Hat production service level agreements (SLAs) and might not be functionally complete. Red{nbsp}Hat does not recommend using them in production. These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.
For more information about the support scope of Red{nbsp}Hat Technology Preview features, see link:https://access.redhat.com/support/offerings/techpreview/[Technology Preview Features Support Scope].
====

The Multiarch Tuning Operator enhances the operational experience within multi-architecture clusters, and single-architecture clusters that are migrating to a multi-architecture compute configuration. 

> This Operator implements the `ClusterPodPlacementConfig` custom resource (CR) to support architecture-aware workload scheduling.

Reviewer 3: ```suggestion
This Operator implements a `ClusterPodPlacementConfig` custom resource (CR) to support architecture-aware workload scheduling.
```
Can we use more than one CR? If so, use "a" instead of "the." Use of the definite/indefinite article is inconsistent throughout. I'd use whichever is accurate throughout.

Reviewer 2: Only one ClusterPodPlacementConfig, named `cluster`, is allowed. Maybe we'd specify this too

Author: Added the following note.

"You can create only a single instance of the `ClusterPodPlacementConfig` CR."

When you create the `ClusterPodPlacementConfig` CR, the Multiarch Tuning Operator deploys an operand. This operand automates the configuration of scheduling predicates for the workloads based on their supported architectures. It uses the `kubernetes.io/arch` label set by the kubelet in the node objects.

> When a pod is created, the operand configures the `nodeAffinity` field in the pod's specification by computing the subset of architectures that the pod can support.

Reviewer 3: ```suggestion
When a pod is created, the operand configures the `nodeAffinity` field in the pod specification by computing the subset of architectures that the pod can support.
```

Before a pod enters the scheduling cycle, the operand adds the `multiarch.openshift.io/scheduling-gate` scheduling gate that prevents the scheduling of the pod until the operand completes the following actions:

. Compute a predicate for the `kubernetes.io/arch` label. 

> . Integrate the predicate as a `nodeAffinity` requirement in the pod's specification.

Reviewer 3: ```suggestion
. Integrate the predicate as a `nodeAffinity` requirement in the pod specification.
```

. Remove the scheduling gate.

> This allows scheduling workloads only on nodes with supported architectures.

Reviewer 3: ```suggestion
These operand actions restrict the scheduling of workloads to those nodes that have the specified supported architectures.
```
I think is what we mean here.

Author: Updated as per the suggestion.

[IMPORTANT]
====

> The Technology Preview version of the Multiarch Tuning Operator does not support clusters with restricted network scenarios.

Reviewer 3: ```suggestion
The Technology Preview version of the Multiarch Tuning Operator does not support clusters with restricted network scenarios.
```
Does "with restricted network scenarios" mean only clusters running on restricted networks, or something broader?

Author: I think it means only clusters running on restricted networks. 

WDYT? @aleskandro

Reviewer 2: it's fine to keep as is for me.

====

> You can install the Multiarch Tuning Operator by using the {product-title} CLI or the {product-title} web console.

Reviewer 3: ```suggestion
```
On the fence about this one. There's no lead-in for uninstalling, and we have uneven TOC flow in this chapter as a whole (L1s for install/uninstall and L2s for the identical modules for creating and deleting). I think you can safely just delete this lead-in. (IMHO it is not valuable to the reader for you to create modules just to level the TOC so it's all identical.)

Author: Deleted this line.
